### PR TITLE
Add support for additional files

### PR DESCRIPTION
**Pull request by mvsusp**
*Issue #, if available:*

*Description of changes:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [X] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [X] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [X] I have updated the [changelog](https://github.com/aws/sagemaker-python-sdk/blob/master/CHANGELOG.rst) with a description of my changes (if appropriate)
- [X] I have updated any necessary [documentation](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/pytorch/estimator.py [hunk 1]
```diff
@@ -47,9 +47,9 @@ def __init__(self, entry_point, source_dir=None, hyperparameters=None, py_versio
         Args:
             entry_point (str): Path (absolute or relative) to the Python source file which should be executed
                 as the entry point to training. This should be compatible with either Python 2.7 or Python 3.5.
-            source_dir (str): Path (absolute or relative) to a directory with any other training
-                source code dependencies aside from tne entry point file (default: None). Structure within this
-                directory are preserved when training on Amazon SageMaker.
+            source_dir (str or [str]): Single path (absolute or relative) or a list of paths to directories with
+                any other training source code dependencies aside from the entry point file (default: None).
+                The structures within this directories are preserved when training on Amazon SageMaker.
```
Comment (iquintero) [inline src/sagemaker/pytorch/estimator.py]: these directories

File: src/sagemaker/chainer/estimator.py [hunk 2]
```diff
@@ -64,9 +64,9 @@ def __init__(self, entry_point, use_mpi=None, num_processes=None, process_slots_
                 set to the number of GPUs on the instance (on GPU instances), or one (on CPU instances).
             additional_mpi_options (str): String of options to the 'mpirun' command used to run the entry point.
                 For example, '-X NCCL_DEBUG=WARN' will pass that option string to the mpirun command.
-            source_dir (str): Path (absolute or relative) to a directory with any other training
-                source code dependencies aside from tne entry point file (default: None). Structure within this
-                directory are preserved when training on Amazon SageMaker.
+            source_dir (str or [str]): Single path (absolute or relative) or a list of paths to directories with
```
Comment (laurenyu) [inline src/sagemaker/chainer/estimator.py]: let's be more explicit and use `list[str]` for lists. also I'd change it to "A single path"

File: src/sagemaker/chainer/estimator.py [hunk 3]
```diff
@@ -64,9 +64,9 @@ def __init__(self, entry_point, use_mpi=None, num_processes=None, process_slots_
...
+                any other training source code dependencies aside from the entry point file (default: None).
```
Comment (laurenyu) [inline src/sagemaker/chainer/estimator.py]: I'd change this line to "any source code (other than the entry point file) needed for training"

File: src/sagemaker/chainer/estimator.py [hunk 4]
```diff
@@ -64,9 +64,9 @@ def __init__(self, entry_point, use_mpi=None, num_processes=None, process_slots_
...
+                The structures within this directories are preserved when training on Amazon SageMaker.
```
Comment (laurenyu) [inline src/sagemaker/chainer/estimator.py]: should we include explanation about the structure of how to access each of the directories? i.e. I assume (but am still reading the PR) that it'll be something like:

```
| base dir from unpacking the tar
| - source dir 1
| - source dir 2
| - etc.
```

based on reading the docstring, but it'd be good to be explicit about it

File: src/sagemaker/fw_utils.py [hunk 5]
```diff
@@ -118,31 +118,43 @@ def tar_and_upload_dir(session, bucket, s3_key_prefix, script, directory):
         bucket (str): S3 bucket to which the compressed file is uploaded.
         s3_key_prefix (str): Prefix for the S3 key.
         script (str): Script filename.
-        directory (str): Directory containing the source file. If it starts with "s3://", no action is taken.
+        directory (str or None): Directory containing the source file. If it starts with "s3://", no action is taken.
```
Comment (laurenyu) [inline src/sagemaker/fw_utils.py]: add `additional_files` to the docstring

File: src/sagemaker/fw_utils.py [hunk 6]
```diff
@@ -118,31 +118,43 @@ def tar_and_upload_dir(session, bucket, s3_key_prefix, script, directory):
...
     """
-    if directory:
-        if directory.lower().startswith("s3://"):
-            return UploadedCode(s3_prefix=directory, script_name=os.path.basename(script))
-        else:
-            script_name = script
-            source_files = [os.path.join(directory, name) for name in os.listdir(directory)]
+    key = '%s/sourcedir.tar.gz' % s3_key_prefix
+
+    if directory and directory.lower().startswith("s3://"):
```
Comment (laurenyu) [inline src/sagemaker/fw_utils.py]: single quotes for the string

File: CHANGELOG.rst [hunk 7]
```diff
@@ -14,6 +14,7 @@ CHANGELOG
 * feature: HyperparameterTuner: Make input channels optional
 * feature: Add support for Chainer 5.0
 * feature: Estimator: add support for MetricDefinitions
+* feature: source_dir accepts a list of directories
```
Comment (jesterhazy) [inline CHANGELOG.rst]: i still haven't heard an explanation for why this is useful. why can't user just stage their files properly in a single source dir before creating an estimator?

Comment (owen-t) [reply CHANGELOG.rst]: From someone on the outside looking in, this looks confusing and unneeded. I understand there's probably a (pressing) reason why we want to do this - but it feels like a hack that we're adding to the PythonSDK to overcome a problem in another system.

Comment (iquintero) [reply CHANGELOG.rst]: I am also against this change because of the same reason Jonathan mentioned.

Comment (leopd) [reply CHANGELOG.rst]: User could stage the files before calling the estimator.  User can also copy their own code to S3, and not use the estimator.  Or they could build their own docker container, and make boto calls instead of using the SDK at all.  The point of this project is to make things easier.

File: src/sagemaker/chainer/estimator.py (hunk 4 above)
Comment (mvsusp) [reply src/sagemaker/chainer/estimator.py]: Yeah, more documentation and examples would be helpful.

File: src/sagemaker/fw_utils.py [hunk 8]
```diff
@@ -118,31 +118,43 @@ def tar_and_upload_dir(session, bucket, s3_key_prefix, script, directory):
...
-    os.remove(tar_file)
 
-    return UploadedCode(s3_prefix='s3://{}/{}'.format(bucket, key), script_name=script_name)
+    try:
+        session.resource('s3').Object(bucket, key).upload_file(tar_file)
+    finally:
+        os.remove(tar_file)
+
+
+def _list_root_files(script, directory, additional_files):
```
[the remark this answers is not in the document]
Comment (mvsusp) [reply src/sagemaker/fw_utils.py]: I am personally against testing private methods, it adds unnecessary coupling to non public facing signatures and does not provide a good grasp of the functionality.

I do agree that testing should be extensive, and cover any edge cases. The tests that I wrote are here https://github.com/aws/sagemaker-python-sdk/pull/494/files#diff-3108f99e19f25f4c77ad4f63d486b174R147

Let me know if you any suggestions of improvement of these methods.
